rad rollback kubernetes command implementation

ytimocin opened the pull request:
# Description

## How to test `rad upgrade kubernetes` and `rad rollback kubernetes`

1. **Binary 1**: Download and install and older version of Radius (0.47 or 0.48).
2. **Binary 2**: Run `make build` on this branch locally and the binary in the dist folder is going to be your 2nd binary for testing.

**Testing Scenarios**

1. Run `rad install` with the first binary and then run `rad version` to see what is installed in the cluster.
2. Try different upgrade scenarios with the second binary. Please pay attention to the help section of the command `rad upgrade kubernetes -h`.
3. Then start playing with the `rad rollback kubernetes` command. Again, please pay attention to the help section of the rollback command.
4. You can combine different scenarios of upgrade and rollback together.
5. Please try different commands like this:
  * `rad rollback kubernetes --revision`
  * `rad rollback kubernetes --list-revisions`
  * `rad upgrade kubernetes`
  * `rad upgrade kubernetes --version X.Y.Z`

## AI-Generated Description of the PR

This pull request introduces a new `rad rollback kubernetes` command to the CLI, enabling users to roll back the Radius control plane on a Kubernetes cluster to a previous revision. The implementation includes command wiring, core logic, test coverage, and required Helm interface extensions.

**Key changes:**

### New Rollback Command Functionality

* Added a new top-level `rad rollback` command and a subcommand `rad rollback kubernetes` that allows users to roll back the Radius installation on a Kubernetes cluster to a previous or specified revision, or list available revisions. (`cmd/rad/cmd/root.go`, `pkg/cli/cmd/rollback/rollback.go`, `pkg/cli/cmd/rollback/kubernetes/kubernetes.go`) [[1]](diffhunk://#diff-d36b7b4d04f01a6cf60c50dd9c01da8fb8c190231ea4c59cbbfbe070178b510bR70-R71) [[2]](diffhunk://#diff-d36b7b4d04f01a6cf60c50dd9c01da8fb8c190231ea4c59cbbfbe070178b510bR388-R393) [[3]](diffhunk://#diff-a7781ef0913086e7f57ee6bf5b3e5e5b8533436212d6182c6b29180cc48d8a67R1-R28) [[4]](diffhunk://#diff-ec6714c4ad06b1011fbc1b4ffc7cf291958b37b88de7ffaa89fd408e0b67dd88R1-R188)

### Helm Interface Enhancements

* Extended the `helm.Interface` with new methods: `RollbackRadius`, `RollbackRadiusToRevision`, and `GetRadiusRevisions`, and added a `RevisionInfo` struct for revision metadata. (`pkg/cli/helm/cluster.go`) [[1]](diffhunk://#diff-a94c29c96fb0778e35d7983669dedfa10de7111040707a3839dace9836b20d0bR158-R175) [[2]](diffhunk://#diff-a94c29c96fb0778e35d7983669dedfa10de7111040707a3839dace9836b20d0bR194-R202)

### Test Coverage

* Added comprehensive unit tests for the new rollback command, covering various scenarios including successful rollbacks, error handling, and revision listing. (`pkg/cli/cmd/rollback/kubernetes/kubernetes_test.go`)

## Type of change
- This pull request adds or changes features of Radius and has an approved issue (issue link required).
Fixes: #issue_number

## Contributor checklist
Please verify that the PR meets the following requirements, where applicable:

- An overview of proposed schema changes is included in a linked GitHub issue.
    - [x] Yes <!-- TaskRadio schema -->
    - [ ] Not applicable <!-- TaskRadio schema -->
- A design document PR is created in the [design-notes repository](https://github.com/radius-project/design-notes/), if new APIs are being introduced.
    - [x] Yes <!-- TaskRadio design-pr -->
    - [ ] Not applicable <!-- TaskRadio design-pr -->
- The design document has been reviewed and approved by Radius maintainers/approvers.
    - [ ] Yes <!-- TaskRadio design-review -->
    - [x] Not applicable <!-- TaskRadio design-review -->
- A PR for the [samples repository](https://github.com/radius-project/samples) is created, if existing samples are affected by the changes in this PR.
    - [ ] Yes <!-- TaskRadio samples-pr -->
    - [x] Not applicable <!-- TaskRadio samples-pr -->
- A PR for the [documentation repository](https://github.com/radius-project/docs) is created, if the changes in this PR affect the documentation or any user facing updates are made.
    - [ ] Yes <!-- TaskRadio docs-pr -->
    - [x] Not applicable <!-- TaskRadio docs-pr -->
- A PR for the [recipes repository](https://github.com/radius-project/recipes) is created, if existing recipes are affected by the changes in this PR.
    - [ ] Yes <!-- TaskRadio recipes-pr -->
    - [x] Not applicable <!-- TaskRadio recipes-pr -->

ytimocin committed:
rad rollback kubernetes command implementation

Signed-off-by: ytimocin <ytimocin@microsoft.com>

brooke-hamilton commented on pkg/cli/cmd/rollback/kubernetes/kubernetes.go:
```diff
@@ -0,0 +1,131 @@
+	}
+
+	commonflags.AddKubeContextFlagVar(cmd, &runner.KubeContext)
+	cmd.Flags().IntVar(&runner.Revision, "revision", 0, "Specify the revision number to rollback to (defaults to previous version)")
```
What does Revision mean? Is it the semantic version? Or does it correspond to a value in Kubernetes? How can I see the available revisions? I think we should update the help text to explain this.

brooke-hamilton reviewed (changes requested):
🚀 

## First issue

I get an error in this scenario. I'm not sure if this is expected or not, but that's because I'm not sure where the previous version would be stored. Maybe there should be an explanation in the help on how Radius determines if a rollback is possible.
1. Install radius using a local build: `make install`
2. `rad install kubernetes`
3. `rad rollback kubernetes`

I thought this would roll back to the previous revision.

```shell
$ rad version
CLI Version Information:
RELEASE   VERSION                  BICEP     COMMIT
edge      v0.47.0-rc1-13-g69031cb  0.36.1    69031cb5403fbb7a371d990aa73ae4b5940689fb

Control Plane Information:
STATUS     VERSION
Installed  0.42.42-dev

$ rad rollback kubernetes
Current Radius version: 0.42.42-dev
Checking for previous revisions...
Error: failed to rollback Radius: no previous revision available for rollback. Current revision count: 1
```

## Second Issue
I did these steps:
1. Install rad CLI version 0.46, and run `rad install kubernetes`.
2. Install local build of rad CLI using `make install`.
3. Run `rad upgrade kubernetes --version latest`.

```shell
$ rad upgrade kubernetes --version latest
Resolved 'latest' to version: v0.47.0
Current Radius version: 0.46.0
Target Radius version: v0.47.0
Running preflight checks...
Running pre-flight checks...
  Running Kubernetes Connectivity...
    ✓ Successfully connected to Kubernetes cluster (version: v1.31.5+k3s1) with sufficient permissions
  Running Helm Connectivity...
    ✓ Helm successfully connected to cluster and found Radius release (version: 0.46.0), Contour dependency found (version: 11.1.1)
  Running Radius Installation...
    ✓ Radius is installed (version: 0.46.0), Contour is installed (version: 11.1.1)
  Running Version Compatibility...
    ✓ Upgrade from 0.46.0 to v0.47.0 is valid
  Running Custom Configuration Validation...
    ✓ No custom configuration parameters provided
  Running Kubernetes Resource Availability...
    ✓ Sufficient resources available for upgrade. Current Radius usage: 0.00 CPU, 570 MB memory. Additional resources needed: 0.00 CPU, 285 MB memory. Available: 8.00 CPU, 15968 MB memory
Pre-flight checks completed successfully
✓ All preflight checks passed
Upgrading Radius from version 0.46.0 to v0.47.0...
Upgrading Radius...
Error: failed to upgrade Radius: failed to prepare Radius Helm chart, err: failed to load Helm chart, err: error downloading helm chart from the registry for version: v0.47.0, release name: radius Cause: failed to perform "FetchReference" on source: ghcr.io/radius-project/helm-chart/radius:v0.47.0: not found.
```

It looks like the rad cli is using the wrong label to pull the helm chart. It is using a label `v0.47.0` when it should be `0.47.0`. See the label [here](https://github.com/radius-project/radius/pkgs/container/helm-chart%2Fradius/422855090?tag=0.47.0). 

This looks like a bug in `rad upgrade` but it is preventing me from testing `rad rollback`.

nicolejms commented:
Hey Yetkin, I don't see the functionality to restore the data that we backed up during upgrade. Is that a miss or did we not add the data backup in upgrade.  I don't think we can release and announce an upgrade story without data backup in case of catestrophic upgrade failure.

ytimocin committed:
Listing revisions, revision 0 behavior documentation, and unit tests refactored to use table-driven testing

Signed-off-by: ytimocin <ytimocin@microsoft.com>

ytimocin commented:
> 🚀
> 
> ## First issue
> I get an error in this scenario. I'm not sure if this is expected or not, but that's because I'm not sure where the previous version would be stored. Maybe there should be an explanation in the help on how Radius determines if a rollback is possible.
> 
> 1. Install radius using a local build: `make install`
> 2. `rad install kubernetes`
> 3. `rad rollback kubernetes`
> 
> I thought this would roll back to the previous revision.
> 
> ```shell
> $ rad version
> CLI Version Information:
> RELEASE   VERSION                  BICEP     COMMIT
> edge      v0.47.0-rc1-13-g69031cb  0.36.1    69031cb5403fbb7a371d990aa73ae4b5940689fb
> 
> Control Plane Information:
> STATUS     VERSION
> Installed  0.42.42-dev
> 
> $ rad rollback kubernetes
> Current Radius version: 0.42.42-dev
> Checking for previous revisions...
> Error: failed to rollback Radius: no previous revision available for rollback. Current revision count: 1
> ```
> 
> ## Second Issue
> I did these steps:
> 
> 1. Install rad CLI version 0.46, and run `rad install kubernetes`.
> 2. Install local build of rad CLI using `make install`.
> 3. Run `rad upgrade kubernetes --version latest`.
> 
> ```shell
> $ rad upgrade kubernetes --version latest
> Resolved 'latest' to version: v0.47.0
> Current Radius version: 0.46.0
> Target Radius version: v0.47.0
> Running preflight checks...
> Running pre-flight checks...
>   Running Kubernetes Connectivity...
>     ✓ Successfully connected to Kubernetes cluster (version: v1.31.5+k3s1) with sufficient permissions
>   Running Helm Connectivity...
>     ✓ Helm successfully connected to cluster and found Radius release (version: 0.46.0), Contour dependency found (version: 11.1.1)
>   Running Radius Installation...
>     ✓ Radius is installed (version: 0.46.0), Contour is installed (version: 11.1.1)
>   Running Version Compatibility...
>     ✓ Upgrade from 0.46.0 to v0.47.0 is valid
>   Running Custom Configuration Validation...
>     ✓ No custom configuration parameters provided
>   Running Kubernetes Resource Availability...
>     ✓ Sufficient resources available for upgrade. Current Radius usage: 0.00 CPU, 570 MB memory. Additional resources needed: 0.00 CPU, 285 MB memory. Available: 8.00 CPU, 15968 MB memory
> Pre-flight checks completed successfully
> ✓ All preflight checks passed
> Upgrading Radius from version 0.46.0 to v0.47.0...
> Upgrading Radius...
> Error: failed to upgrade Radius: failed to prepare Radius Helm chart, err: failed to load Helm chart, err: error downloading helm chart from the registry for version: v0.47.0, release name: radius Cause: failed to perform "FetchReference" on source: ghcr.io/radius-project/helm-chart/radius:v0.47.0: not found.
> ```
> 
> It looks like the rad cli is using the wrong label to pull the helm chart. It is using a label `v0.47.0` when it should be `0.47.0`. See the label [here](https://github.com/radius-project/radius/pkgs/container/helm-chart%2Fradius/422855090?tag=0.47.0).
> 
> This looks like a bug in `rad upgrade` but it is preventing me from testing `rad rollback`.

Addressed the issues. I will do one final review tomorrow morning but here is my feedback for the issues you mentioned above:

**Issue 1**: This is expected - you need at least two revisions to rollback. The first installation creates revision 1, so there's nothing to rollback to yet. You need to upgrade first, then you can rollback.

**Issue 2**: You're correct. The Helm chart tags don't include the 'v' prefix. Use: `rad upgrade kubernetes --version 0.47.0`

brooke-hamilton reviewed (approved):
🚀 🚢 Looks good to me. The code is in good shape. I was able to test several scenarios. Here's one of them: 
```shell
$ ./dist/linux_amd64/release/rad rollback kubernetes --list-revisions
Current Radius version: 0.49.0
REVISION  CHART VERSION  STATUS      UPDATED              DESCRIPTION
4         0.49.0         deployed    2025-08-06 10:16:29  Upgrade complete
3         0.48.0         superseded  2025-08-06 10:16:29  Rollback to 1
2         0.49.0         superseded  2025-08-06 10:16:29  Upgrade complete
1         0.48.0         superseded  2025-08-06 10:16:29  Install complete

$ ./dist/linux_amd64/release/rad rollback kubernetes --revision 1
Current Radius version: 0.49.0
Rolling back to specified revision 1...
Rolling back Radius to revision 1 (version 0.48.0)...
Radius rollback completed successfully!
✓ Radius rollback completed successfully!

$ ./dist/linux_amd64/release/rad rollback kubernetes --list-revisions
Current Radius version: 0.48.0
REVISION  CHART VERSION  STATUS      UPDATED              DESCRIPTION
5         0.48.0         deployed    2025-08-06 10:16:29  Rollback to 1
4         0.49.0         superseded  2025-08-06 10:16:29  Upgrade complete
3         0.48.0         superseded  2025-08-06 10:16:29  Rollback to 1
2         0.49.0         superseded  2025-08-06 10:16:29  Upgrade complete
1         0.48.0         superseded  2025-08-06 10:16:29  Install complete
```
